src/v/raft/seq_stm.cc

```diff
@@ -0,0 +1,337 @@
+      "seq",
+      ss::default_priority_class())
+  , _log(logger)
+  , _config(config) {}
+
+ss::future<>
+seq_stm::hydrate_snapshot([[maybe_unused]] storage::snapshot_reader& reader) {
+    return reader.read_metadata().then([this, &reader](iobuf meta_buf) {
+        iobuf_parser meta_parser(std::move(meta_buf));
+        seq_snapshot_header hdr;
+        hdr.version = reflection::adl<int8_t>{}.from(meta_parser);
+        vassert(
+          hdr.version == seq_snapshot_header::supported_version,
+          "unsupported seq_snapshot_header version {}",
+          hdr.version);
+        hdr.snapshot_size = reflection::adl<int32_t>{}.from(meta_parser);
+
+        return read_iobuf_exactly(reader.input(), hdr.snapshot_size)
+          .then([this](iobuf data_buf) {
+              iobuf_parser data_parser(std::move(data_buf));
+              auto data = reflection::adl<snapshot>{}.from(data_parser);
+              for (auto& entry : data.entries) {
+                  auto pid_seq = _seq_table.find(entry.pid);
+                  if (pid_seq == _seq_table.end()) {
+                      _seq_table.emplace(entry.pid, entry);
+                  } else if (pid_seq->second.seq < entry.seq) {
+                      pid_seq->second = entry;
+                  }
+              }
+              _last_snapshot_offset = data.offset;
+              _insync_offset = data.offset;
+              vlog(
+                raftlog.trace,
+                "seq_stm snapshot with offset:{} is loaded",
+                data.offset);
+          });
+    });
+}
+
+ss::future<> seq_stm::persist_snapshot(iobuf&& data) {
+    seq_snapshot_header header;
+    header.version = seq_snapshot_header::supported_version;
+    header.snapshot_size = data.size_bytes();
+
+    iobuf data_size_buf;
+    reflection::serialize(data_size_buf, header.version, header.snapshot_size);
+
+    return _snapshot_mgr.start_snapshot().then(
+      [this, data = std::move(data), data_size_buf = std::move(data_size_buf)](
+        storage::snapshot_writer writer) mutable {
+          return ss::do_with(
+            std::move(writer),
+            [this,
+             data = std::move(data),
+             data_size_buf = std::move(data_size_buf)](
+              storage::snapshot_writer& writer) mutable {
+                return writer.write_metadata(std::move(data_size_buf))
+                  .then([&writer, data = std::move(data)]() mutable {
+                      return write_iobuf_to_output_stream(
+                        std::move(data), writer.output());
+                  })
+                  .finally([&writer] { return writer.close(); })
+                  .then([this, &writer] {
+                      return _snapshot_mgr.finish_snapshot(writer);
+                  });
+            });
+      });
+}
+
+ss::future<> seq_stm::make_snapshot() {
+    return _op_lock.with([this]() {
+        auto f = ss::now();
+        if (!_resolved_when_snapshot_hydrated.available()) {
```

Comment:
maybe we can make the follwing 3 lines of code a method f.e. `wait_for_snapshot_hydrated()`. Then the code would look like

```c++

return wait_for_snapshot_hydrated().then(...

```

```